transports/websys/src/websocket.rs

```diff
@@ -0,0 +1,321 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+//! Libp2p websocket transports built on [Websys](https://rustwasm.github.io/wasm-bindgen/web-sys/index.html).
+use futures::{future::Ready, io, prelude::*};
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerId, TransportError, TransportEvent},
+};
+use parking_lot::Mutex;
+use send_wrapper::SendWrapper;
+use wasm_bindgen::{prelude::*, JsCast};
+use web_sys::{MessageEvent, WebSocket};
+
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    sync::Arc,
+    task::Poll,
+    task::{Context, Waker},
+};
+
+/** A Websocket transport that can be used in a Wasm client.
+
+## Example
+
+To create an authenticated transport instance with Noise protocol and Mplex:
+
+```no_run
+# use libp2p_core::{upgrade::Version, Transport};
+# use libp2p_identity::Keypair;
+# use libp2p_mplex::MplexConfig;
+# use libp2p_noise::NoiseAuthenticated;
+# use libp2p_websys_transport::websocket;
+let local_key = Keypair::generate_ed25519();
+let transport = websocket::Transport::default()
+    .upgrade(Version::V1)
+    .authenticate(NoiseAuthenticated::xx(&local_key).unwrap())
+    .multiplex(MplexConfig::default())
+    .boxed();
+```
+*/
+#[derive(Default)]
+pub struct Transport;
+
+impl libp2p_core::Transport for Transport {
+    type Output = Connection;
+    type Error = Error;
+    type ListenerUpgrade = Ready<Result<Self::Output, Self::Error>>;
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(&mut self, _addr: Multiaddr) -> Result<ListenerId, TransportError<Self::Error>> {
+        Err(TransportError::Other(Error::NotSupported))
+    }
+
+    fn remove_listener(&mut self, _id: ListenerId) -> bool {
+        false
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let ws_url = if let Some(url) = websocket_url(addr) {
+            url
+        } else {
+            return Err(TransportError::Other(Error::NotSupported));
+        };
+
+        Ok(async move {
+            let socket = match WebSocket::new(&ws_url) {
+                Ok(ws) => ws,
+                Err(err) => return Err(Error::JsError(format!("{err:?}"))),
```

Comment:
I am not sure debug printing will display something nice here. Is this `JsValue` always a string? Can we just print that string instead?

Reply:
I'll check if we can get a string.